tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,209 @@
+print(output)
+# True
+```
+
+First of all, we are creating a function that takes two integer values, sums them up and returns the output of an incoming callback. This callback `is_positive` is defined as a function that takes one mandatory integer and one optional string. If the message exists it will print it, and always returns if the incoming value is positive. Of course, this function is not pure because of the print side-effect, but it’s just an example :)
+
+## Generics and Union Types
+
+One of the most powerful features in a type system is the generics and union types. It’s also possible to use these in Python annotations. Let’s take a look at the following example:
+
+```python
+from typing import TypeVar, List
+
+
+T = TypeVar("T", int, List[str])
+
+def generic_add(x: T, y: T) -> T:
+   return x + y
+
+
+x1: int = 5
+y1: int = 2
+print(generic_add(x1, y1))
+# 7
+
+x2: List[str] = ["Hello"]
+y2: List[str] = ["World"]
+print(generic_add(x2, y2))
+# ['Hello', 'World']
+
+x3: str = "foo"
+y3: str = "bar"
+print(generic_add(x3, y3))
+# error: Value of type variable "T" of "generic_add" cannot be "str"
+```
+
+Above, we defined a type variable `T` that can be integer or List of strings and a `generic_add` function that perform the “addition” operation for the incoming arguments which are limited by the `TypeVar` declaration.
+
+The first two invocations will work because we're passing arguments that belong to the `TypeVar` set of types and the function will behave accordingly since both types implements the `+` operation. It will actually run for the third invocation because Python implements the add operation for strings, but mypy will raise the error as `str` is not supported by the `T` definition.
+
+Also, we can make use of union types. Let’s suppose we want to support any number that can only be of types integer and float:
+
+```python
+from typing import Union
+
+
+Number = Union[float, int]
+
+def union_add(x: Number, y: Number) -> Number:
+   return x + y
+
+
+x1: int = 5
+y1: int = 2
+print(union_add(x1, y1))
+# 7
+
+x2: float = 3.14
+y2: float = 3.14
+print(union_add(x2, y2))
+# 6.28
+
+x3: str = "2"
+y3: str = "1"
+print(union_add(x3, y3))
+# error: Argument 1 to "union_add" has incompatible type "str"; expected "Union[float, int]"
+# error: Argument 2 to "union_add" has incompatible type "str"; expected "Union[float, int]"
+```
+
+To wrap up, static type checking will always be important to prevent bugs or implementation misuse, and allows us to take one of the main advantages of _Type-Driven Development_, don't rely too much in unit tests that check the values consistency like other dynamic and weak-typing languages do. It's not about stop testing, but leave them the work of testing something specific and, in the same way, complement our projects with this set of tools. Now you can start doing type checking automatically in your local environments or as a Continuous Integration job along with the rest of your pipeline.
```

Comment:
After misuse maybe a new sentence? It allows us to take advantage of Type-Driven Development without relying on an excessive amount of unit tests that check the values. How are we defining weak-typing languages? Is this what we're calling "dynamically-typed" languages?